### PR TITLE
z_select decompiled, 1 non matching left

### PR DESCRIPTION
**Pull request by fig02**
This file was started by maide https://github.com/zeldaret/oot/pull/146
I picked up where they left off, cleaned things up, and matched all but 1 function.

Other things:
-VIEWPORT_INIT changed to INIT_FULLSCREEN_VIEWPORT based on findings in this file
-GfxPrint_Ctor and GfxPrint_Dtor renamed to GfxPrint_Init and GfxPrint_Destroy

### CODE REVIEW

File: spec [hunk 1]
```diff
@@ -559,9 +559,8 @@ endseg
 beginseg
     name "ovl_select"
     include "build/src/overlays/gamestates/ovl_select/z_select.o"
-    include "build/data/overlays/gamestates/z_select.data.o"
-    include "build/data/overlays/gamestates/z_select.rodata.o"
     include "build/data/overlays/gamestates/z_select.reloc.o"
+    //include "build/src/overlays/gamestates/ovl_select/ovl_select_reloc.o"
```
Comment (mzxrules) [inline spec:563]: what's this?

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 2]
```diff
@@ -1,25 +1,633 @@
...
+        }
+
+        if (CHECK_PAD(pad1->press, U_JPAD)) {
+            if (this->unk_22C == 1) {
+                this->unk_224 = 0;
+            }
+            if (this->unk_224 == 0) {
+                this->unk_224 = 0x14;
+                this->unk_22C = 1;
+                Audio_PlaySoundGeneral(0x1800, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
```
Comment (mzxrules) [inline src/overlays/gamestates/ovl_select/z_select.c:279]: shouldn't 0x1800 be `NA_SE_IT_SWORD_IMPACT`, from sfx.h?

File: spec (hunk 1 above)
Comment (fig02) [reply spec:563]: its the auto generated reloc. it cant be used right now because of the non matching

File: include/z64.h [hunk 3]
```diff
@@ -1136,6 +1136,40 @@ typedef struct {
     /* 0x01E2 */ char unk_1E2[0x06];
 } TitleContext; // size = 0x1E8
 
+typedef struct {
+    /* 0x00 */ char* name;
+    /* 0x04 */ void (*loadFunc)(struct SelectContext*, s32);
+    /* 0x08 */ s32 entranceIndex;
+} MapEntry; // size = 0xC
```
Comment (fig02) [inline include/z64.h:1143]: just want to note im changing any name with map in it to scene before this gets merged

Comment (Roman971) [reply include/z64.h:1143]: Noted. I suggest `SceneSelectEntry` for this particular struct, so it's not confused with other "scene" structs like scene table entries (called `Scene` currently).

Comment (fig02) [reply include/z64.h:1143]: agreed. ill change stuff in a min

Comment (fig02) [reply include/z64.h]: changes made

File: include/z64.h [hunk 4]
```diff
@@ -188,7 +188,7 @@ typedef struct {
     /* 0x13F6 */ s16          unk_13F6;
     /* 0x13F8 */ s16          unk_13F8;
     /* 0x13FA */ u16          eventInf[4]; // "event_inf"
-    /* 0x1402 */ u16          mapIndex; // intended for maps/minimaps but commonly used as the dungeon index
+    /* 0x1402 */ u16          mapIndex; // intended for scenes/minimaps but commonly used as the dungeon index
```
Comment (Roman971) [inline include/z64.h:191]:
```suggestion
    /* 0x1402 */ u16          mapIndex; // intended for maps/minimaps but commonly used as the dungeon index
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 5]
```diff
@@ -1,25 +1,633 @@
...
+    gSaveContext.buttonStatus[4] = BTN_ENABLED;
+    gSaveContext.buttonStatus[3] = BTN_ENABLED;
+    gSaveContext.buttonStatus[2] = BTN_ENABLED;
+    gSaveContext.buttonStatus[1] = BTN_ENABLED;
+    gSaveContext.buttonStatus[0] = BTN_ENABLED;
+    gSaveContext.unk_13E7 = gSaveContext.unk_13E8 = gSaveContext.unk_13EA = gSaveContext.unk_13EC = 0;
+    Audio_SetBGM(NA_BGM_STOP);
+    gSaveContext.entranceIndex = entranceIndex;
+    gSaveContext.respawnFlag = 0;
+    gSaveContext.respawn[0].entranceIndex = -1;
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:37]:
```suggestion
    gSaveContext.respawn[RESPAWN_MODE_DOWN].entranceIndex = -1;
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 6]
```diff
@@ -1,25 +1,633 @@
...
+    gSaveContext.nightSeqIndex = 0xFF;
+    gSaveContext.unk_13C7 = 1;
+    D_8011FB30 = 0;
+    this->state.running = false;
+    SET_NEXT_GAMESTATE(&this->state, Gameplay_Init, GlobalContext)
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/gamestates/ovl_select/func_808017F0.s")
+static SceneSelectEntry sScenes[] = {
+    { " 1:SPOT00", Select_LoadGame, 0x000000CD },
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:47]: Entrance indexes are normally written as shorts, not words.
```suggestion
    { " 1:SPOT00", Select_LoadGame, 0x00CD },
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 7]
```diff
@@ -1,25 +1,633 @@
...
+    { "122:Sutaru", Select_LoadGame, 0x00000047 },
+    { "123:jikkenjyou", Select_LoadGame, 0x000002EA },
+    { "124:depth\x8Cﾃｽﾄ", Select_LoadGame, 0x000000B6 },
+    { "125:\x8Cﾊｲﾗﾙ\x8Dﾆﾜ\x8Cｹﾞｰﾑ2", Select_LoadGame, 0x00000076 },
+    { "title", Select_LoadTitle, 0x00000000 },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/gamestates/ovl_select/func_8080187C.s")
+//mostly regalloc, a good amount of instruction ordering. confirmed equivalent in game.
+#ifdef NON_MATCHING
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:176]:
```suggestion
#ifdef NON_MATCHING
// mostly regalloc, a good amount of instruction ordering. confirmed equivalent in game.
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 8]
```diff
@@ -1,25 +1,633 @@
...
+    "\x8Dｼﾊﾞﾗｸｵﾏﾁｸﾀﾞｻｲ", // "Please wait a minute"
+    "\x8Dﾁｮｯﾄ ﾏｯﾃﾈ",     // "Hold on a sec"
+    "\x8Cｳｪｲﾄ ｱ ﾓｰﾒﾝﾄ",  // "Wait a moment"
+    "\x8Cﾛｰﾄﾞ\x8Dﾁｭｳ",   // "Loading"
+    "\x8Dﾅｳ ﾜｰｷﾝｸﾞ",     // "Now working"
+    "\x8Dｲﾏ ﾂｸｯﾃﾏｽ",     // "Now creating"
+    "\x8Dｺｼｮｳｼﾞｬﾅｲﾖ",    // "It's not broken"
+    "\x8Cｺｰﾋｰ ﾌﾞﾚｲｸ",    // "Coffee Break"
+    "\x8C"
+    "Bﾒﾝｦｾｯﾄｼﾃｸﾀﾞｻｲ",                      // "Please set B side"
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:419]:
```suggestion
    "\x8C\Bﾒﾝｦｾｯﾄｼﾃｸﾀﾞｻｲ",                      // "Please set B side"
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 9]
```diff
@@ -1,25 +1,633 @@
...
+    GfxPrint_SetColor(printer, 255, 255, 255, 255);
+    randomMsg = Math_Rand_ZeroOne() * ARRAY_COUNT(sLoadingMessages);
+    GfxPrint_Printf(printer, "%s", sLoadingMessages[randomMsg]);
+}
+
+static char* sAgeLabels[] = {
+    "\x8D"
+    "17(ﾜｶﾓﾉ)", // "17(young)"
+    "\x8D"
+    "5(ﾜｶｽｷﾞ)", // "5(very young)"
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:438]:
```suggestion
    "\x8D\17(ﾜｶﾓﾉ)", // "17(young)"
    "\x8D\5(ﾜｶｽｷﾞ)", // "5(very young)"
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 10]
```diff
@@ -1,25 +1,633 @@
...
+
+void Select_DrawLoadingScreen(SelectContext* this) {
+    GraphicsContext* gfxCtx;
+    GfxPrint* printer;
+    Gfx* dispRefs[5];
+
+    gfxCtx = this->state.gfxCtx;
+    Graph_OpenDisps(dispRefs, gfxCtx, "../z_select.c", 977);
+
+    gSPSegment(gfxCtx->polyOpa.p++, 0x00, 0x00000000);
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:544]:
```suggestion
    gSPSegment(gfxCtx->polyOpa.p++, 0x00, NULL);
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 11]
```diff
@@ -1,25 +1,633 @@
...
+}
+
+void Select_Draw(SelectContext* this) {
+    GraphicsContext* gfxCtx;
+    Gfx* dispRefs[5];
+
+    gfxCtx = this->state.gfxCtx;
+    Graph_OpenDisps(dispRefs, gfxCtx, "../z_select.c", 1013);
+    
+    gSPSegment(gfxCtx->polyOpa.p++, 0x00, 0x00000000);
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:566]:
```suggestion
    gSPSegment(gfxCtx->polyOpa.p++, 0x00, NULL);
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 12]
```diff
@@ -1,25 +1,633 @@
...
+    func_80095248(gfxCtx, 0, 0, 0);
+    SET_FULLSCREEN_VIEWPORT(&this->view)
+    func_800AAA50(&this->view, 0xF);
+
+    if (this->state.running == false) {
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:571]:
```suggestion
    if (!this->state.running) {
```

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 13]
```diff
@@ -1,25 +1,633 @@
...
+    Select_PrintMenu(this, printer);
+    {
+        s32 arg;
+
+        arg = gSaveContext.linkAge;
+        Select_PrintAgeSetting(this, printer, arg);
+
+        arg = gSaveContext.cutsceneIndex & 0xFFFF;
+        Select_PrintCutsceneSetting(this, printer, arg);
+    }
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:529]: Is the `arg` stuff here really required? That looks like a fake match to me.

File: src/overlays/gamestates/ovl_select/z_select.c [hunk 14]
```diff
@@ -1,25 +1,633 @@
...
+
+void Select_DrawMenu(SelectContext* this) {
+    GraphicsContext* gfxCtx;
+    GfxPrint* printer;
+    Gfx* dispRefs[5];
+
+    gfxCtx = this->state.gfxCtx;
+    Graph_OpenDisps(dispRefs, gfxCtx, "../z_select.c", 930);
+    
+    gSPSegment(gfxCtx->polyOpa.p++, 0x00, 0x00000000);
```
Comment (Roman971) [inline src/overlays/gamestates/ovl_select/z_select.c:512]:
```suggestion
    gSPSegment(gfxCtx->polyOpa.p++, 0x00, NULL);
```

File: src/overlays/gamestates/ovl_select/z_select.c (hunk 9 above)
Comment (fig02) [reply src/overlays/gamestates/ovl_select/z_select.c:438]: this doesnt work, or the one above

Comment (fig02) [reply src/overlays/gamestates/ovl_select/z_select.c:438]: actually it works for the one above but not this one?? lol

**Pull request merged**